### DISCUSSION TITLE
File

### DISCUSSION THREAD

**Discussion by ux-fran** [nl-design-system/backlog#347] [Component suggestions]
<img width="480" alt="visuele weergave van een file component" src="https://github.com/user-attachments/assets/b9be0fb5-13d5-43ad-9814-59e624625542" />

We willen graag meer leren over de File component. In welke vorm, of vormen, wordt deze momenteel ingezet en toegepast binnen jouw organisatie?

Deel hieronder een link naar Figma, Storybook, Github of screenshots. Met deze informatie kunnen we bewijzen dat de component onderdeel moet worden van de NL Design System component library.

Daarnaast wordt het delen van onderzoek gewaardeerd. Dus heb je zelf onderzoek gedaan en zijn er inzichten voor dit component? Dan horen we dit ook graag!

[Link naar bericht in Slack](https://codefornl.slack.com/archives/C7KDQRXMZ/p1729179387703749)

---

## Naam

File

## Beschrijving

Toont meta-informatie over een bestand en biedt één of meer acties, zoals bekijken, downloaden of verwijderen.

## Ook bekend als

Download, Uploads

## Gerelateerde componenten

[File Input](https://github.com/orgs/nl-design-system/discussions/310)

## Gerelateerde patronen

[Bestanden toevoegen - Meerstappenformulier](https://github.com/orgs/nl-design-system/discussions/381)

## Onderzoek

[Rapport Generieke Services 2023 - Gemeente Den Haag (PDF, 9 MB)](https://github.com/user-attachments/files/16993296/Rapport.generieke.services.23032023.def.pdf)
Uit dit gebruikersonderzoek is er één inzicht over de documenten die op een zaakpagina staan:

Een respondent verwachtte de zojuist geüploade informatie of foto’s terug te vinden onder ‘Documenten’. Het kan nuttig zijn om duidelijker te communiceren wie de afzender van de documenten is, aangezien deze sectie ook bestanden kan bevatten die door de gemeente zijn geüpload.

[Vraag toegankelijkheid en focus - Slack](https://codefornl.slack.com/archives/C7KDQRXMZ/p1677576902552619)
Rozerin stelde de vraag waar de de focus moet komen te liggen bij de File component. Moet deze komen bij de gehele component, enkel op de Link of zelfs op meerdere elementen die niet zijn benoemd? De community antwoordt op deze vraag in de thread. 

Conclusie: bij de gemeente Den Haag is besloten de focus te leggen op de gehele component in plaats van alleen de link, omdat het gehele element aanklikbaar is en niet alleen de link.

[Should “Download” components be buttons or links? - Accessible Web](https://accessibleweb.com/question-answer/should-download-components-be-buttons-or-links/)
Vraag of een download component een Button of een Link moet zijn op het gebied van toegankelijkheid. De aanbeveling is: een Link.

[File uploader - Carbon Design System](https://www.carbondesignsystem.com/components/file-uploader/usage/)
Carbon Design biedt functionaliteit voor het downloaden, bekijken en verwijderen van bestanden. Voorvertoning van bestanden is beschikbaar voordat ze worden gedownload.

[File List - Ant Design](https://ant.design/components/upload/#components-upload-demo-fileList)
Ant Design heeft een bestandslijstcomponent waarmee gebruikers bestanden kunnen vooraf bekijken, downloaden en verwijderen via acties in een gepresenteerde lijst.

[Files - Lightning Design System](https://www.lightningdesignsystem.com/components/files/)
Salesforce biedt een component voor het beheren van bestanden waarmee gebruikers deze kunnen downloaden, bekijken en verwijderen, ontworpen voor consistentie en toegankelijkheid.

[File-upload - The Component Gallery](https://component.gallery/components/file-upload/)
Een groot overzicht van alle File (input/upload) componenten van verschillende design systems.

**Comment by ux-fran** [comment]
## Research nov 2024
-  Gekeken in diverse files van Amsterdam (ik heb geen toegang tot de Figma), DSO, RVO en Utrecht en daar verder iets vergelijkbaars gevonden.

Heb ik wat gemist of weet jij (de lezer) misschien nog een organisatie of andere gemeente met een vergelijkbaar component, vul gerust aan!

**Comment by ux-fran** [comment]
## Den Haag

- [Storybook](https://nl-design-system.github.io/denhaag/?path=/docs/react-surfaces-files--docs)
- [Figma - Huidig](https://www.figma.com/design/JpoY3waVoQGlLQzQXTL9nn/%F0%9F%8C%BB-Den-Haag-Design-System?node-id=10817-61351&t=uz0TWRngCXR1NFzv-4)
- [GitHub](https://github.com/nl-design-system/denhaag/tree/main/components/File)

---

### Screenshots / Varianten

**Storybook**

Default
<img width="700" alt="image" src="https://github.com/user-attachments/assets/b8a10194-e935-4e35-92aa-c3ac029292b2" />

Hover
<img width="700" alt="image" src="https://github.com/user-attachments/assets/63949e9a-034e-43ed-87c8-671ac788cd10" />

Focus
<img width="700" alt="image" src="https://github.com/user-attachments/assets/4edd0d01-22df-45e1-981d-4ba48ce83f48" />

Button
<img width="700" alt="image" src="https://github.com/user-attachments/assets/7b66c511-23fa-414e-a82f-18b133e7dd9a" />

List
<img width="700" alt="image" src="https://github.com/user-attachments/assets/f93089c1-f203-43f5-92da-8fa3c297ddad" />

Responsive gedrag
<img width="300" alt="image" src="https://github.com/user-attachments/assets/ae8e256d-cc8a-486a-83f3-a0c811a3ca9e" />

**Figma**

Als file-uploader
<img width="700" alt="image" src="https://github.com/user-attachments/assets/771a7536-9da4-4018-91a7-f639f9b33861" />


---


### Notities

- Bestaat uit de classes `left` en `right`. In `left` staat het icoon dat het bestandstype aanduidt, en in `right` bevinden zich het label en de link.
- Het gehele element is aanklikbaar.
- Bij hover verandert alleen de styling van de link mee.
- Label bestaat uit 'name' en 'description'. De description bevat het bestandstype, de bestandsgrootte en de datum van upload of beschikbaarheid.
- Kan als een lijst worden getoond, waarbij de bestanden onder elkaar worden gestapeld.
- Responsive gedrag: Link wordt op een volgende regel geplaatst.
- In Figma is de File component ook zichtbaar wanneer deze wordt geüpload via een File-uploader component. Het bestand kan desgewenst worden verwijderd.
- Richtlijnen beschikbaar.

**Comment by ux-fran** [comment]
## DUO

[Documentatie website](https://uno.dfront.rijkscloud.nl/#/components/file)

---

### Screenshots / Varianten
Default
<img width="600" alt="image" src="https://github.com/user-attachments/assets/d089cb80-3659-4ea4-beb9-0f6e2d5d69cc" /><br>

File
Om bijvoorbeeld te gebruiken in een tabel.
<img width="300" alt="image" src="https://github.com/user-attachments/assets/0a16b56d-bce8-4209-8ca6-468a455a7c43" /><br>

Downloadbaar bestand
Gebruik bij een bestand wat te downloaden is.
<img width="300" alt="image" src="https://github.com/user-attachments/assets/18072b6e-54d8-48f9-95d9-3c292be8c2f0" />

Responsive gedrag
<img width="300" alt="image" src="https://github.com/user-attachments/assets/997d2fe4-ef08-451e-b4ac-cfdb1f7146c3" />

---

### Notities

- Wordt gebruikt wanneer de gebruiker één of meerdere bestanden wil uploaden of moet downloaden.
- Bevat standaard een icoon van een bestand.
- Type default: bevat een Link (verwijderen). Alleen deze is aanklikbaar. Het gehele element is dus **niet** aanklikbaar.
- Type file: bevat geen Link. 
- Type downloadbaar bestand: bestandsnaam is een Link.
- Filename bestaat uit 'title' en 'size'. De title geeft de naam van het bestand aan. En de size geeft de bestandsextensie en bestandsgrootte aan.
- Bestandsgrootte wordt aangegeven in KB's en MB's.
- Kan als een lijst worden getoond, waarbij de bestanden onder elkaar worden gestapeld.
- Responsive gedrag: filename en Link worden onder elkaar geplaatst.
- Richtlijnen beschikbaar.

**Comment by ux-fran** [comment]
## Open Formulieren

[Storybook](https://open-formulieren.github.io/open-forms-sdk/?path=/story/form-io-components-vanilla-fileupload--happy-flow})

---

### Screenshots / Varianten

Happy flow
<img width="772" alt="image" src="https://github.com/user-attachments/assets/1059ced5-79a2-45ef-a9ad-b1b7857f1b98" />

One allowed file type
<img width="761" alt="image" src="https://github.com/user-attachments/assets/b7881c48-bd9b-4e39-87cb-e6278a6a210e" />

Multiple allowed file types
<img width="766" alt="image" src="https://github.com/user-attachments/assets/06a435e1-cb7d-4e2c-a59a-868cc45d81e2" />

Error flow
<img width="761" alt="image" src="https://github.com/user-attachments/assets/d670b913-574c-4423-b84d-90326ef728b0" />

Responsive gedrag
<img width="300" alt="image" src="https://github.com/user-attachments/assets/cea370f1-2bbc-4193-9496-a78b4cae73a5" />

---

### Notities

- Er is geen afzonderlijk ‘File’ component beschikbaar; voorbeelden maken deel uit van de ‘File-upload’ component.
- Wordt weergegeven wanneer de gebruiker één of meerdere bestanden heeft geüpload.
- Bevat een bestandsnaam, een verwijdericoon, de bestandsgrootte en, bij fouten (unhappy flows), een foutmelding.
- Interactie verwijdericoon: geüploade bestand wordt direct verwijderd uit het overzicht.
- In de happy flow is de bestandsnaam een klikbare link.
- Bestandsgrootte wordt aangegeven in kB's.
- Het voorbeeld ondersteunt het uploaden van maximaal één bestand.
- Responsive gedrag: de layout blijft consistent, ongeacht het schermformaat.

**Comment by ux-fran** [comment]
## Open Gemeenten

[Patternlab](https://patternlab.lenteveld.nl/patterns/02-organisms-200-Miscellaneous-30-Uploads-Uploads/02-organisms-200-Miscellaneous-30-Uploads-Uploads.html)

---

### Screenshots / Varianten

<img width="700" alt="image" src="https://github.com/user-attachments/assets/9674d9fb-dfb8-4aa4-a77b-3b2bea152d44" />

---


### Notities

- Onder de naam: Uploads.
- Is in zijn geheel een Link.
- Bestaat uit een icoon, het soort bestand, bestandsextensie en bestandsgrootte.
- Icoon geeft bestandstype aan. Er zijn verschillende iconen beschikbaar.
- Bestandsgrootte wordt aangegeven in MB's.
- Kan als een lijst worden getoond, waarbij de bestanden onder elkaar worden gestapeld.
- Responsive gedrag: de layout blijft consistent, ongeacht het schermformaat.

**Comment by ux-fran** [comment]
## Rotterdam

[Figma](https://www.figma.com/design/iWuQsxelJzXVfCge87Og81/RODS-component?node-id=1-6723&t=kl7gR2ar7KwqZW5V-4)

---


### Screenshots / Varianten

<img width="455" alt="image" src="https://github.com/user-attachments/assets/61c62ffc-9a3b-4e75-ab01-f5368db20084" />

Responsive gedrag
<img width="286" alt="image" src="https://github.com/user-attachments/assets/a297175d-27fd-418c-9f06-d2e7969583d7" />


---


### Notities

- Er is geen afzonderlijk ‘File’ component beschikbaar; voorbeelden maken deel uit van de ‘Modal file upload’ component.
- Wordt weergegeven wanneer de gebruiker één of meerdere bestanden heeft geüpload.
- Bevat een icoon, bestandsnaam, de bestandsgrootte en een kruisicoon. 
- Bestandsgrootte wordt aangegeven in KB's.
- Bij meerdere geüploade bestanden worden deze onder elkaar gestapeld.
- Responsive gedrag: de layout blijft consistent, ongeacht het schermformaat. Geen voorbeeld uitgewerkt bij langere bestandsnamen.

**Comment by jiromaykin** [comment]
## Open Inwoner

Nog niet omgezet naar NL Design System.
Is nu nog een Django HTML File component (dus nog niet in Storybook)

- [Figma](https://www.figma.com/design/iKGhWhstaLIlFSaND2q7cE/OIP---Designs--new-?node-id=9268-28875&m=dev) (Zaak detailpagina)
- [GitHub](https://github.com/maykinmedia/open-inwoner/blob/develop/src/open_inwoner/components/templates/components/File/File.html) (Django HTML)

---

## Afbeeldingen

### Lijst van geüploade Files, in 'Mijn Zaken'

Een document dat minder dan 24 uur geleden is geüpload, krijgt een 'nieuw' label met rode notificatie badge, en een rode border-left.

<img width="579" height="188" alt="bestand-lijst-OIP" src="https://github.com/user-attachments/assets/b5c41e3a-280f-445c-850f-e308e8880667" />

### Één 'Plan' bestand in 'Mijn Samenwerkingen' 

Omdat daar maximaal slechts 1 werk plan in mag staan,
oudere versies van plan-sjablonen worden in een File list weergegeven die er anders uit ziet qua styling:

<img width="801" height="294" alt="Bestand-oudnieuw-OIP" src="https://github.com/user-attachments/assets/c45bec3f-56c4-4c73-8b3a-3a3396c54bfc" />

### Opties voor andere soorten iconen

Afhankelijk van waar File gebruikt wordt:
als het bestand in de 'file upload' fase is en er eerst foutmeldingen zichtbaar moeten zijn, voor het geval er een te groot of verkeerd type bestand wordt geüpload door gebruiker, moet er de mogelijkheid worden geboden om het geselecteerde File (dat dus nog niet is geüpload) te kunnen verwijderen, d.m.v. Delete icoon,
ook kan er afhankelijk van File type een adner soort icoon worden getoond.

<img width="903" height="529" alt="bestand-UploadDelete-OIP" src="https://github.com/user-attachments/assets/28e2630d-7828-449d-9432-8ba876e88d55" />

**Reply by jiromaykin** [reply]
De File component wordt nu geïmplementeerd in Open-Inwoner met gebruik van die van Den Haag. Dus deze wordt binnkrot beschikbaar in onze Storybook.
Ook zal hier feedback toegevoegd worden over de zaken waar we tegenaan liepen en wat er verbeterd zou kunnen worden aan de Den Haag File versie.

Bijvoorbeeld: in de Den Haag versie is de gehele File component een link, maar dat zou eigenlijk niet heel netjes zijn voor screenreaders, die dan alles gaan oplezen zonder onderscheid. In de versie van OIP wordt de download 'button' de enige link, en in het geval dat die button wordt vervangen door een 'verwijderen' button, wordt dit een echte button waar javascript (met een delete functie) aan gekoppeld mag worden (beter geen javascript functie koppelen aan een `<a>` link).

**Comment by Rozerinay** [comment]
## Amsterdam

- [Storybook](https://designsystem.amsterdam/?path=/docs/components-forms-file-list--docs)
- [GitHub](https://github.com/Amsterdam/design-system/tree/develop/packages/css/src/components/file-list)
- [Figma](https://www.figma.com/design/9IGm6IdPUYizBNGsUnueBd/Amsterdam-Design-System?node-id=36431-17477&t=3tDnLkyEA5NQAyIP-11)

---

### Screenshots / Varianten

**Storybook**

<img width="560" height="166" alt="image" src="https://github.com/user-attachments/assets/df4435e4-e109-432d-b07f-e5f9890465e5" />

Lange bestandsnamen
Dit is een eigen test om te kijken hoe de component omgaat met lange bestandsnamen.
<img width="584" height="170" alt="image" src="https://github.com/user-attachments/assets/3873f7f6-65a4-46ec-9dde-5a9d205a605c" />

Afbeelding preview
Bij het uploaden van een afbeelding, wordt in plaats van een icoon, bij `preview`, een voorvertoning van de afbeelding getoond.
<img width="563" height="84" alt="image" src="https://github.com/user-attachments/assets/24e15879-0dce-4549-b851-fbd0cb415198" />

---

### Notities
- Er is geen afzonderlijk ‘File’ component beschikbaar; voorbeelden maken deel uit van de ‘File List’ component door middel van `__item`.
- Bevat een icoon of afbeelding (`preview`), bestandsnaam, extensie en grootte (`info`) en een verwijder button.
- Bestandsgrootte wordt aangegeven in bytes, kB's en MB's. 
- Bij meerdere geüploade bestanden worden deze onder elkaar gestapeld.
- Responsive gedrag: de layout blijft consistent, ongeacht het schermformaat. Geen voorbeeld uitgewerkt bij langere bestandsnamen, maar wel nagebootst in code.

**Comment by Rozerinay** [comment]
## DSO

- [Storybook](https://storybook.dso-toolkit.nl/88.0.0/?path=/story/html-css-form-group--files)
- [GitHub](https://github.com/dso-toolkit/dso-toolkit/tree/master/packages/dso-toolkit/src/components/form-group/files)

---

### Screenshots / Varianten

**Storybook**

Large Container
<img width="1052" height="236" alt="image" src="https://github.com/user-attachments/assets/f8421496-a17c-46ca-894f-95e7dae7c81f" />

Small Container
<img width="394" height="359" alt="image" src="https://github.com/user-attachments/assets/57d22d2b-e900-4d0d-9bf0-19147002b848" />

Lange bestandsnamen
Dit is een eigen test om te kijken hoe de component omgaat met lange bestandsnamen.
<img width="391" height="384" alt="image" src="https://github.com/user-attachments/assets/d5556a1d-9c7e-4ff5-b6c5-9a4ccb938c61" />

<img width="1019" height="261" alt="image" src="https://github.com/user-attachments/assets/4a93c585-4877-4709-b87b-9d1c86b283b4" />

---

### Notities
- Er is geen afzonderlijk ‘File’ component beschikbaar; voorbeelden maken deel uit van de Form Group.
- Bevat de bestandsnaam, download button en verwijder button.
- Kan verder ook een Checkbox, Spinner en een icoon bevatten die een status weergeeft.
- Bij meerdere geüploade bestanden worden deze onder elkaar gestapeld.
- Responsive gedrag: de layout past zich aan.
- Geen voorbeeld uitgewerkt bij langere bestandsnamen, maar wel nagebootst in code.

**Comment by Rozerinay** [comment]
## ✨ Dit component is nu Help Wanted ✨

Help je mee hem door de Community stappen te krijgen?
[File op het Community bord](https://github.com/orgs/nl-design-system/projects/29/views/1?filterQuery=file+348)

**Comment by jiromaykin** [comment]
Aanvulling vanuit **Open Inwoner** nu het component van Den Haag daarin is gebruikt in productie:

- File component van Den Haag [in Open-Inwoner Storybook](https://maykinmedia.github.io/open-inwoner/?path=/docs/components-file--docs)
- [Github code van het component](https://github.com/maykinmedia/open-inwoner/blob/develop/src/open_inwoner/react/components/File/File.tsx)
- [Github met de aangepaste SCSS](https://github.com/maykinmedia/open-inwoner/blob/develop/src/open_inwoner/react/components/File/File.scss)

---

**Wat OIP van Den Haag File component gebruikt:**

CSS/HTML classes en styling (.denhaag-file, .denhaag-file__left, etc.)

**Eisen van OIP  / verschillen in implementatie**

Belangrijkste: In de versie van OIP wordt de download 'button' de enige `<a>` link, en in het geval dat die button wordt vervangen door een 'verwijderen' button, wordt dit een echte `<Button>` waar javascript (met een delete functie) aan gekoppeld mag worden (beter geen javascript functie koppelen aan een <a> link

1. van de hele File bar heeft OIP een Div gemaakt in plaats van een `<a>` link (beter voor schermlezers en specifiekere feedback op de download/delete functies)
2. specifieke styling voor de download-link toegevoegd (text-decoration none)
3. Custom Material Icons (in plaats van SVG)
4. Delete button met eigen functionaliteit en styling toevoegen en/of overschrijven (.denhaag-link--remove de border weghalen)
5. OIP gebruikt nog [een 3e variant zonder delete of download functie](https://www.figma.com/design/iKGhWhstaLIlFSaND2q7cE/OIP---Designs--new-?node-id=9733-33049&t=6dEwbWpuBc5YJJuy-1), alleen bedoeld voor preview (bij kiezen voor Samenwerkings plan sjabloon)
6. styling voor de uppercase file extensie en bestandsgrootte
7. missende design-token voor border-radius van het hele File component.

**Comment by Rozerinay** [comment]
## ✨ Dit component is nu Community ✨

Help je mee hem door de Candidate stappen te krijgen?
[File op het Candidate bord](https://github.com/orgs/nl-design-system/projects/80/views/1?filterQuery=file+348)

**Comment by UXTeamGroningen** [comment]
Bij ons hebben we nog aanvullingen over het file component in Mijn Omgeving / OIP op basis van onderzoek:
https://gebruikersonderzoeken.nl/docs/onderzoek-bekijken/mijn-zaken/groningen-mijn-omgeving/

Door ons zaaksysteem worden automatisch al een aantal bestanden toegevoegd aan de detailpagina. Namelijk dat bevestigingsmail en de aanvraag zelf. Hierdoor loopt het aantal documenten dat wordt toegevoegd aan een zaak al wel snel op.

Bevindingen in het onderzoek waren:
- Ook als een inwoner dan documenten up moet loaden, krijg je al snel een soort van 'bak' aan documenten, onderin de zaak.
- Die bak maakt het onduidelijk om te weten welk document afkomstig is van wie (gemeente of inwoner zelf)
- Op basis van het bovenstaande hebben we in ieder geval gekozen voor een indicator die onderscheid aangeeft in nieuwe en oude documenten
